### PR TITLE
Refactor ValidateConfigInfo and delete some redundant judgment

### PR DESCRIPTION
**Pull request by ping035627**
What type of PR is this?
/kind cleanup

What this PR does / why we need it:
Refactor ValidateConfigInfo to reduce its complexity, and delete some redundant judgment.

Which issue(s) this PR fixes:

Fixes #

Special notes for your reviewer:

Does this PR introduce a user-facing change?:

/release-note-none
/priority backlog

### CODE REVIEW

**Comment by neolit123** [comment]
@kubernetes/sig-cluster-lifecycle-pr-reviews

**Comment by SataQiu** [comment]
Thansk @ping035627 
/ok-to-test

**Comment by ping035627** [comment]
hi, test has passed, can you review again? thanks.

**Comment by ping035627** [comment]
@neolit123 can you review this pr? thanks.

File: cmd/kubeadm/app/discovery/file/file.go [hunk 1]
```diff
@@ -140,6 +102,54 @@ func ValidateConfigInfo(config *clientcmdapi.Config, clustername string) (*clien
...
+	)
+
+	// load pre-existing client certificates
+	if config.Contexts[config.CurrentContext] != nil && len(config.AuthInfos) > 0 {
+		user := config.Contexts[config.CurrentContext].AuthInfo
+		authInfo, ok := config.AuthInfos[user]
+		if !ok || authInfo == nil {
+			return nil, errors.Errorf("empty settings for user %q", user)
+		}
+		if len(authInfo.ClientCertificate) != 0 {
```
Comment (ping035627) [inline cmd/kubeadm/app/discovery/file/file.go]: validateKubeConfig has ensured that len(authInfo.ClientKeyData)  must be 0 when len(authInfo.ClientCertificate) != 0, so here needn't to judge whether len(authInfo.ClientKeyData) is zero or not.

File: cmd/kubeadm/app/discovery/file/file.go [hunk 2]
```diff
@@ -140,6 +102,54 @@ func ValidateConfigInfo(config *clientcmdapi.Config, clustername string) (*clien
...
+			clientCert, err := ioutil.ReadFile(authInfo.ClientCertificate)
+			if err != nil {
+				return nil, err
+			}
+			authInfo.ClientCertificateData = clientCert
+		}
+		if len(authInfo.ClientKey) != 0 {
```
Comment (ping035627) [inline cmd/kubeadm/app/discovery/file/file.go]: the same to above.

File: cmd/kubeadm/app/discovery/file/file.go (hunk 1 above)
Comment (ereslibre) [reply cmd/kubeadm/app/discovery/file/file.go]: I might be missing something but based on `clientcmd.Validate` what I see is that it excludes `ClientKey` and `ClientKeyData` from both being provided (or none), as well as `ClientCertificate` and `ClientCertificateData`, but I don't see this relationship that `len(authInfo.ClientCertificate) != 0 => len(authinfo.ClientKeyData) == 0`.

Comment (ping035627) [reply cmd/kubeadm/app/discovery/file/file.go]: Sorry, there are typos in my comment, it should be following:
validateKubeConfig has ensured that len(authInfo.ClientCertificateData) must be 0 when len(authInfo.ClientCertificate) != 0, so here needn't to judge whether len(authInfo.ClientCertificateData) is zero or not.

if len(authInfo.ClientCertificateData) == 0 && len(authInfo.ClientCertificate) != 0 

I delete "len(authInfo.ClientCertificateData) == 0".

File: cmd/kubeadm/app/discovery/file/file.go [hunk 3]
```diff
@@ -140,6 +102,54 @@ func ValidateConfigInfo(config *clientcmdapi.Config, clustername string) (*clien
 	return refreshedBaseKubeConfig, nil
 }
 
+// get kubeconfig from config
+func getKubeconfig(config *clientcmdapi.Config, defaultCluster *clientcmdapi.Cluster, clustername string) (*clientcmdapi.Config, error) {
```
Comment (fabriziopandini) [inline cmd/kubeadm/app/discovery/file/file.go]: IMO default ClusterConfiguration should not be passed as a parameter
It is derived from config (that is already a parameter) and never used outside of this function

**Review by fabriziopandini — CHANGES_REQUESTED** [review]
@ping035627 Thanks for this PR!
I'm in favour of breaking down big function int smaller ones, but in this case I must admit that I don't understand what is the real boundary/split of responsibility between `ValidateConfigInfo` and the new function `getKubeconfig`; might be a better name and some more details in the comment can help here.

One other thing that could add value to this refactor, is the fact that once a responsibility is splitted from the original function and assigned to `getKubeconfig`, unit test are provided to ensure that this behaviour is preserved over time.

File: cmd/kubeadm/app/discovery/file/file.go (hunk 3 above)
Comment (ping035627) [reply cmd/kubeadm/app/discovery/file/file.go]: The new function getKubeconfig is extracts the part of create kubeconfig from ValidateConfigInfo. I have added the unit test for getKubeconfig.

**Review by rosti — COMMENTED** [review]
Thanks @ping035627 !
I am not quite sure about this one. I see the point of splitting it, but I also see the point of @fabriziopandini.
We need a better name and a more clear contract for what the func should be doing.

**Comment by ping035627** [comment]
@rosti I think the cyclomatic complexity of ValidateConfigInfo is too high, we can extract the logic of create kubeconfig to a independent function. Maybe the function named "createKubeconfig" is more appropriate.

**Pull request closed without merging**